sdk/go/common/resource/properties_path.go

```diff
@@ -299,6 +301,184 @@ func (p PropertyPath) Contains(other PropertyPath) bool {
 	return true
 }
 
+func (p PropertyPath) reset(old, new PropertyValue) bool {
+	if len(p) == 0 {
+		return false
+	}
+
+	// If this is the last component we want to do the reset, else we want to search for the next component.
+	key := p[0]
+	switch key := key.(type) {
+	case int:
+		// An index < 0 is always a path error, even for empty arrays or objects
+		if key < 0 {
+			return false
+		}
+
+		// This is a leaf path element, so we want to reset the value at this index in new to the value at this index from old
+		if len(p) == 1 {
+			if !old.IsArray() && !new.IsArray() {
+				// Neither old nor new are arrays, so we can't reset this index
+				return true
+			} else if !old.IsArray() || !new.IsArray() {
+				// One of old or new is an array but the other isn't, so this is a path error
+				return false
+			}
+
+			// If neither array contains this index then this is a _same_ and so ok, e.g. given old:[1, 2] and
+			// new:[1] and a path of [3] we can return true because new at [3] is the same as old at [3], it
+			// doesn't exist.
+			if key >= len(old.ArrayValue()) && key >= len(new.ArrayValue()) {
+				return true
+			}
+			// If one array has this index but the other doesn't this is a path failure because we can't
+			// remove a location from an array.
+			if key >= len(old.ArrayValue()) || key >= len(new.ArrayValue()) {
+				return false
+			}
+			// Otherwise both arrays contain this index and we can reset the value of it in new to what is in
+			// old.
+			v := old.ArrayValue()[key]
+			new.ArrayValue()[key] = v
+			return true
+		}
+
+		if !old.IsArray() || !new.IsArray() {
+			// At least one of old or new is not an array, so we can't keep searching along this path but
+			// we only return an error if both are not arrays.
+			return !old.IsArray() && !new.IsArray()
+		}
+
+		// If this index is out of bounds in either array then this is a path failure because we can't
+		// continue the search of this path down each PropertyValue.
+		if key >= len(old.ArrayValue()) || key >= len(new.ArrayValue()) {
+			return false
+		}
+		old = old.ArrayValue()[key]
+		new = old.ArrayValue()[key]
+		return p[1:].reset(old, new)
```

Comment:
These lines are not covered by the Reset tests. Would help to have non-leaf array items in the Reset unit tests.

Reply:
I'll add some more tests to cover these and the other ones you've pointed out, good catches I was pretty sure I'd missed some but was going a bit tunnel vision looking at it!